runtime/common/src/parachains.rs

```diff
@@ -142,6 +231,145 @@ pub enum Origin {
 	Parachain(ParaId),
 }
 
+/// An offence that is filed if the validator has submitted a double vote.
+#[derive(RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq, Eq))]
+pub struct DoubleVoteOffence<Offender> {
+	/// The current session index in which we report a validator.
+	session_index: SessionIndex,
+	/// The size of the validator set in current session/era.
+	validator_set_count: u32,
+	/// An offender that has submitted two conflicting votes.
+	offender: Offender,
+}
+
+impl<Offender: Clone> Offence<Offender> for DoubleVoteOffence<Offender> {
+	const ID: Kind = *b"double-vote:doub";
+	type TimeSlot = SessionIndex;
+
+	fn offenders(&self) -> Vec<Offender> {
+		vec![self.offender.clone()]
+	}
+
+	fn session_index(&self) -> SessionIndex {
+		self.session_index
+	}
+
+	fn validator_set_count(&self) -> u32 {
+		self.validator_set_count
+	}
+
+	fn time_slot(&self) -> Self::TimeSlot {
+		self.session_index
+	}
+
+	fn slash_fraction(_offenders_count: u32, _validator_set_count: u32) -> Perbill {
+		// Slash 100%.
+		Perbill::from_percent(100)
+	}
+}
+
+impl<O: Clone> ParachainsOffence<O> for DoubleVoteOffence<O> {
+	fn new(
+		session_index: SessionIndex,
+		validator_set_count: u32,
+		offender: O,
+	) -> Self {
+		Self {
+			session_index,
+			validator_set_count,
+			offender,
+		}
+	}
+}
+
+pub trait ParachainsOffence<O>: Offence<O> {
+	fn new(
+		session_index: SessionIndex,
+		validator_set_count: u32,
+		offender: O,
+	) -> Self;
+}
+
+pub trait HandleDoubleVote<T: Trait> {
+	/// The proof of key ownership.
+	type Offence: ParachainsOffence<IdentificationTuple<T>>;
+
+	fn check_report(
+		report: &DoubleVoteReport,
+		key_owner_proof: Proof,
+		i: Vec<u8>
+	) -> Option<IdentificationTuple<T>>;
+
+	fn report_offence(
+		reporters: Vec<T::AccountId>,
+		offence: Self::Offence
+	);
+}
+
+impl<T: Trait> HandleDoubleVote<T> for () {
+	type Offence = DoubleVoteOffence<IdentificationTuple<T>>;
+
+	fn check_report(
+		_report: &DoubleVoteReport,
+		_key_owner_proof: Proof,
+		_i: Vec<u8>,
+	) -> Option<IdentificationTuple<T>> {
+		None
+	}
+
+	// TODO: As soon as substrate is updated and `report_offence` returns
+	// a result, this function should also be changed to returning a result.
+	fn report_offence(
+		_reporters: Vec<T::AccountId>,
+		_offence: Self::Offence,
+	) {}
+}
+
+pub struct DoubleVoteHandler<P, R, O> {
+	_phantom: rstd::marker::PhantomData<(P, R, O)>,
+}
+
+impl<P, R, O> Default for DoubleVoteHandler<P, R, O> {
+	fn default() -> Self {
+		Self {
+			_phantom: Default::default(),
+		}
+	}
+}
+
+impl<T, P, R, O> HandleDoubleVote<T> for DoubleVoteHandler<P, R, O>
+where
+	T: Trait<HandleDoubleVote = Self>,
+	P: KeyOwnerProofSystem<
+		// AccountId?
+		(sp_core::crypto::KeyTypeId, Vec<u8>),
```

Comment:
Does the `AccountId?` refer to the `Vec<u8>` here?

I think this should rather be the `ValidatorId` type and the `KeyTypeId` should be `PARACHAIN_KEY_TYPE_ID`. These are part of the session keys that validators submit to the chain each session.

This is the flow of what would happen:
  - A validator submits their session keys. The session module perceives each pair of session keys roughly as a `(KeyTypeId, Vec<u8>)`. A session starts with those keys, and validators use them to sign messages while they are actice.
  - The session ends and the historical module creates an entry `(KeyTypeId, Vec<u8>) -> IdentificationTuple` in a trie for each key for each validator. The `Vec<u8>` is the key data and it uniquely identifies the validator in that context.
  - We want to come along later and prove entries out of the trie. So we need to pass `PARACHAIN_KEY_TYPE_ID` and the `ValidatorId` of the offender.

Reply:
So, yes this is one of the pain points I was talking about: the `historical` module implementation of `KeyOwnerProofSystem` is generic over `(KeyTypeId, D)` where `D: AsRef<[u8]>` and is a serialized value of type `T::AccountId`, I assume.

It is true that in main module code `T::AccountId = ValidatorId` (if I am not mistaken), but, for instance in testing `T::AccountId = u64`. So it looks like we can not boldly use the exact type, here, rather

1. Be generic over it with some trait bounds 
2. Just go ahead and use it in an encoded form, as it is done now.

Reply:
> It is true that in main module code T::AccountId = ValidatorId (if I am not mistaken), but, for instance in testing T::AccountId = u64. So it looks like we can not boldly use the exact type, here, rather

The type of `AccountId` should not be relevant here, only the type of `SessionKey`s.